File: backend/app/Models/ShareLink.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\MorphTo;
use Illuminate\Support\Facades\DB;

/**
 * Polymorphic share link — file atau folder bisa punya banyak link,
 * masing-masing dengan expiry & max_views sendiri. Coexists dengan
 * legacy `files.share_token` & `folders.share_token` di tabel files/folders;
 * lihat ShareLink::resolveActive() dan FileController::viewByToken.
 *
 * Status state machine:
 *   - revoked_at NULL  + expires_at NULL/future + views < max → ACTIVE
 *   - revoked_at set   → REVOKED (manual)
 *   - expires_at past  → EXPIRED (auto, via ExpireShareLinksJob)
 *   - views >= max     → EXHAUSTED (auto, dicek runtime di resolveActive)
 */
#[Fillable([
    'user_id',
    'shareable_type',
    'shareable_id',
    'token',
    'expires_at',
    'max_views',
    'views_count',
    'revoked_at',
])]
class ShareLink extends Model
{
    use HasUuids;

    protected $keyType = 'string';
    public $incrementing = false;

    protected function casts(): array
    {
        return [
            'expires_at' => 'datetime',
            'revoked_at' => 'datetime',
            'max_views' => 'integer',
            'views_count' => 'integer',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function shareable(): MorphTo
    {
        return $this->morphTo();
    }

    /**
     * Filter ke row yang masih aktif: belum di-revoke, belum expired,
     * dan belum capai max_views. Untuk dipakai di query list endpoint.
     */
    public function scopeActive(Builder $query): Builder
    {
        return $query
            ->whereNull('revoked_at')
            ->where(function (Builder $q) {
                $q->whereNull('expires_at')->orWhere('expires_at', '>', now());
            })
            ->where(function (Builder $q) {
                $q->whereNull('max_views')
                    ->orWhereColumn('views_count', '<', 'max_views');
            });
    }

    /**
     * Lookup by token, validate state, dan increment view counter secara
     * atomic (race-safe). Return null kalau token tidak ada atau sudah
     * tidak valid — caller treat sebagai 410 Gone.
     *
     * Increment dilakukan via raw DB::table()->increment() supaya tidak
     * race dengan concurrent request ke token yang sama saat counter
     * sudah di ambang max_views.
     */
    public static function resolveActive(string $token): ?self
    {
        $row = DB::table('share_links')
            ->where('token', $token)
            ->whereNull('revoked_at')
            ->where(function ($q) {
                $q->whereNull('expires_at')->orWhere('expires_at', '>', now());
            })
            // Cegah increment lewat max_views: kalau max_views NULL
            // atau views_count masih < max_views, boleh lewat.
            ->where(function ($q) {
                $q->whereNull('max_views')
                    ->orWhereColumn('views_count', '<', 'max_views');
            })
            ->first();

        if (! $row) {
            return null;
        }

        // Atomic increment — kalau proses lain sudah bikin lewat max
        // saat race, query di bawah akan null (return 410). Aman.
        $updated = DB::table('share_links')
            ->where('id', $row->id)
            ->where(function ($q) {
                $q->whereNull('max_views')
                    ->orWhereColumn('views_count', '<', 'max_views');
            })
            ->increment('views_count');

        if ($updated === 0) {
            return null; // race: lewat max_views di antara read & increment
        }

        return static::find($row->id);
    }
}

File: mobile/lib/data/models/share_link.dart
/// Mirrors backend `ShareLinkResource`. Polymorphic share link dengan
/// expiry & max_views — coexist dengan `share_token` legacy di
/// FileItem/Folder.
///
/// Status state machine (lihat backend ShareLink model):
///   - revoked_at != null        → REVOKED
///   - expires_at di masa lalu   → EXPIRED
///   - views_count >= max_views  → EXHAUSTED
///   - else                      → ACTIVE
class ShareLink {
  const ShareLink({
    required this.id,
    required this.token,
    required this.url,
    required this.previewUrl,
    required this.viewsCount,
    required this.isActive,
    required this.shareableType,
    required this.shareableId,
    required this.createdAt,
    this.expiresAt,
    this.maxViews,
    this.revokedAt,
  });

  final String id;
  final String token;
  final String url;
  final String previewUrl;
  final DateTime? expiresAt;
  final int? maxViews;
  final int viewsCount;
  final DateTime? revokedAt;
  final bool isActive;
  final String shareableType;
  final String shareableId;
  final DateTime createdAt;

  /// Computed: `is_active` di backend boleh stale (race antara
  /// resolveActive dan storage), jadi cek ulang di client.
  bool get currentlyActive {
    if (revokedAt != null) return false;
    if (expiresAt != null && expiresAt!.isBefore(DateTime.now())) return false;
    if (maxViews != null && viewsCount >= maxViews!) return false;
    return isActive;
  }

  factory ShareLink.fromJson(Map<String, dynamic> json) => ShareLink(
        id: (json['id'] ?? '').toString(),
        token: (json['token'] ?? '') as String,
        url: (json['url'] ?? '') as String,
        previewUrl: (json['preview_url'] ?? '') as String,
        expiresAt: _parseDate(json['expires_at']),
        maxViews: json['max_views'] as int?,
        viewsCount: (json['views_count'] ?? 0) as int,
        revokedAt: _parseDate(json['revoked_at']),
        isActive: (json['is_active'] ?? false) as bool,
        shareableType: (json['shareable_type'] ?? '') as String,
        shareableId: (json['shareable_id'] ?? '').toString(),
        createdAt: _parseDate(json['created_at']) ?? DateTime.now(),
      );

  static DateTime? _parseDate(dynamic raw) {
    if (raw == null) return null;
    if (raw is String && raw.isNotEmpty) return DateTime.tryParse(raw);
    return null;
  }
}